# Aggregate Functions and Axis-Based Math (Summarizing Data)

In software development, we often need to summarize arrays of numbers. For example, in Angular/TypeScript,
if you want to find the sum or average of an array, you might write:

```typescript
const total = numbers.reduce((sum, num) => sum + num, 0);
const average = total / numbers.length;
```

This is fine for small arrays. But in data science, you might have arrays with millions of numbers. Writing loops or `.reduce()` in plain Python is very slow because Python is an interpreted language.

NumPy provides **Aggregate Functions** (functions that take a whole array of numbers and summarize them into a single value, like a sum, mean, or maximum). Under the hood, these functions are written in pre-compiled C-code, making them incredibly fast.

More importantly, NumPy allows you to calculate these summaries **along specific axes**. If you have a 2D matrix (a grid of rows and columns), you can find the sum of each row individually, or the average of each column, in a single line of code!


### Simple Explanation with an Everyday Analogy
Imagine a spreadsheet grid of student test scores:
*   Each **row** represents a single student.
*   Each **column** represents a specific subject (Math, Science, English).

If you want to summarize this data, you have three choices:
1.  **No Axis specified**: You sum up every single number in the entire spreadsheet. This gives you the total points scored by all students in all subjects combined.
2.  **Axis 0 (Columns / Vertical)**: You squash the grid vertically. You sum each column. This gives you the class total score for Math, Science, and English individually.
3.  **Axis 1 (Rows / Horizontal)**: You squash the grid horizontally. You sum each row. This gives you the total score for each individual student.


#### Basic Aggregations on a 1D Array
Let's see the most common statistical functions:
*   `np.sum(a)`: Calculates the sum of all elements.
*   `np.mean(a)`: Calculates the arithmetic average.
*   `np.median(a)`: Calculates the middle value when the array is sorted.
*   `np.std(a)`: Calculates the standard deviation (how spread out the numbers are).
*   `np.min(a)` and `np.max(a)`: Find the smallest and largest values.
*   `np.argmin(a)` and `np.argmax(a)`: Return the **index** of the smallest and largest values (instead of the value itself).

In [1]:
import numpy as np

a = np.array([40, 10, 50, 30, 20])

print("Array a:", a)
print("Sum:", np.sum(a))
print("Mean (Average):", np.mean(a))
print("Median:", np.median(a))
print("Standard Deviation:", np.round(np.std(a), 2))
print("Minimum value:", np.min(a))
print("Index of minimum value (np.argmin):", np.argmin(a))
print("Maximum value:", np.max(a))
print("Index of maximum value (np.argmax):", np.argmax(a))

Array a: [40 10 50 30 20]
Sum: 150
Mean (Average): 30.0
Median: 30.0
Standard Deviation: 14.14
Minimum value: 10
Index of minimum value (np.argmin): 1
Maximum value: 50
Index of maximum value (np.argmax): 2


#### Multi-dimensional Aggregations and the "Axis" Parameter
In a 2D array:
*   `axis=0` means perform the operation **along the rows** (vertically, column by column).
*   `axis=1` means perform the operation **along the columns** (horizontally, row by row).

*A quick rule of thumb:* When you specify an axis, that is the dimension that gets **collapsed** (or squashed).
If you have a `(3, 4)` shape array and you sum with `axis=0`, you "collapse" the 3 rows, leaving you with a 1D array of shape `(4,)`.

In [2]:
import numpy as np

matrix = np.array([
    [1, 2, 3],
    [4, 5, 6]
])

print("Original 2D matrix:\n", matrix)

# 1. Total sum (no axis specified)
print("Total sum:", np.sum(matrix))

# 2. Sum along Axis 0 (collapse rows, sum columns vertically)
print("Sum of columns (axis=0):", np.sum(matrix, axis=0))

# 3. Sum along Axis 1 (collapse columns, sum rows horizontally)
print("Sum of rows (axis=1):", np.sum(matrix, axis=1))

Original 2D matrix:
 [[1 2 3]
 [4 5 6]]
Total sum: 21
Sum of columns (axis=0): [5 7 9]
Sum of rows (axis=1): [ 6 15]


### Common Mistakes Beginners Make
1.  **Confusing Axis 0 and Axis 1**: Many beginners think `axis=0` means rows, so they expect `np.sum(matrix, axis=0)` to sum each row. **It is the opposite!** `axis=0` sums *vertically* down the columns, and `axis=1` sums *horizontally* across the rows. Remember: the axis you specify is the one that is destroyed/collapsed.
2.  **Using `argmin` or `argmax` expecting row/column coordinates on 2D arrays**: By default, if you do `np.argmax(matrix)` on a 2D array, NumPy flattens the array in its head and returns a single index number. If you want row/col coordinates, you must use `np.unravel_index(np.argmax(matrix), matrix.shape)` or specify an `axis` parameter!



#### Exercise 1
You have a 1D array representing the daily temperature in Celsius for a week:
```python
temps = np.array([19.5, 21.0, 18.0, 25.5, 20.0, 22.5, 19.0])
```
1. Find the average (mean) temperature.
2. Find the warmest day's temperature (max).
3. Find which day of the week was the coldest. (Get its index using `np.argmin`).

In [3]:
import numpy as np
temps = np.array([19.5, 21.0, 18.0, 25.5, 20.0, 22.5, 19.0])

mean_temp = np.mean(temps)
max_temp = np.max(temps)
coldest_day_idx = np.argmin(temps)

print("Average Temp:", np.round(mean_temp, 2))
print("Warmest Temp:", max_temp)
print("Coldest Day Index (0 = Mon):", coldest_day_idx)

Average Temp: 20.79
Warmest Temp: 25.5
Coldest Day Index (0 = Mon): 2


*Explanation*: Index `2` (Wednesday, if we start counting on Monday at index 0) is the coldest day with `18.0` degrees.



#### Exercise 2
You run a small business. You have a 2D array of sales where each row is a different store, and each column is a quarter (Q1, Q2, Q3, Q4):
```python
sales = np.array([
    [10000, 12000, 15000, 18000],  # Store A
    [8000,  9500,  11000, 13000],  # Store B
    [15000, 14000, 16000, 21000]   # Store C
])
```
1. Calculate the total sales for *each store* across the year (use the correct axis!).
2. Calculate the average sales for *each quarter* across all stores.

In [4]:
import numpy as np
sales = np.array([
    [10000, 12000, 15000, 18000],
    [8000,  9500,  11000, 13000],
    [15000, 14000, 16000, 21000]
])

# 1. Total sales per store (sum horizontally across columns)
store_totals = np.sum(sales, axis=1)

# 2. Average sales per quarter (mean vertically down rows)
quarter_averages = np.mean(sales, axis=0)

print("Total sales per store (Store A, B, C):", store_totals)
print("Average sales per quarter (Q1, Q2, Q3, Q4):", quarter_averages)

Total sales per store (Store A, B, C): [55000 41500 66000]
Average sales per quarter (Q1, Q2, Q3, Q4): [11000.         11833.33333333 14000.         17333.33333333]


*Explanation*:
*   `axis=1` squashes columns. It adds numbers across the columns: e.g., `10000 + 12000 + 15000 + 18000 = 55000` for Store A.
*   `axis=0` squashes rows. It averages columns vertically: e.g., for Q1, `(10000 + 8000 + 15000) / 3 = 11000.0`.



#### Exercise 3
Given a 3D array representing a sensor reading over 2 different days, where each day has a 3x3 grid of sensor values:
```python
sensors = np.array([
    # Day 1 sensor readings
    [[10, 12, 11],
     [14, 15, 13],
     [11, 10, 12]],
    # Day 2 sensor readings
    [[11, 13, 11],
     [16, 14, 12],
     [12, 11, 13]]
])
```
Compute the average reading for each sensor cell across the two days. (The output should be a single 3x3 matrix showing the average of Day 1 and Day 2 for each cell). Think about which axis to collapse!

In [5]:
import numpy as np
sensors = np.array([
    [[10, 12, 11],
     [14, 15, 13],
     [11, 10, 12]],

    [[11, 13, 11],
     [16, 14, 12],
     [12, 11, 13]]
])

# We want to collapse the "days" axis (which is Axis 0, the first dimension)
# Shape of sensors is (2, 3, 3) -> 2 days, 3 rows, 3 columns.
# Collapsing Axis 0 leaves us with a (3, 3) matrix.
average_readings = np.mean(sensors, axis=0)
print(average_readings)

[[10.5 12.5 11. ]
 [15.  14.5 12.5]
 [11.5 10.5 12.5]]


*Explanation*: NumPy looks at position `[0, 0]` across both days: `10` on Day 1 and `11` on Day 2. The average is `(10 + 11) / 2 = 10.5`. It does this automatically for every single coordinate.
